<launch>
    <node pkg="tf" type="static_transform_publisher" name="lidar_broadcaster" args="0.0 0 0.278189004081027 0 0 0 1 base_link os_sensor 100" />
    <node pkg="tf" type="static_transform_publisher" name="imu_broadcaster" args="0.0 0 0.278189004081027 0 0 0 1 base_link ins_base_link 100" />
    <node pkg="tf" type="static_transform_publisher" name="gps_broadcaster" args="0 0 0 0.7071068, 0, 0, 0.7071068 base_link ins_base_link 100" />

    <node pkg="robot_localization" type="ekf_localization_node" name="ekf_odom_node" output="screen" >

        <!-- Basic parameters -->
        <param name="frequency" value="30"/>
        <param name="sensor_timeout" value="0.1"/>
        <param name="two_d_mode" value="false"/>
        <remap from="odometry/filtered" to="odom/ekf/enc_imu"/>
        <param name="transform_time_offset" value="0.0"/>

        <!-- Frames -->
        <param name="map_frame" value="map"/>
        <param name="odom_frame" value="odom"/>
        <param name="base_link_frame" value="base_link"/>
        <param name="world_frame" value="map"/>

        <!-- Lidar Odometry -->
        <param name="odom0" value="/aft_mapped_to_init"/>
        <param name="odom0_differential" value="false" />
        <param name="odom0_relative" value="false" />
        <param name="odom0_queue_size" value="10" />
        <rosparam param="odom0_config">[true,  true, true,
                                        false, false, false,
                                        true, true, false,
                                        false, false, true,
                                        false, false, false]</rosparam>

        <!-- Standalone IMU Config-->
        <param name="imu0" value="/imu"/>
        <param name="imu0_differential" value="false" />
        <param name="imu0_relative" value="true" />
        <param name="imu0_queue_size" value="10" />
        <param name="imu0_remove_gravitational_acceleration" value="true" />
        <rosparam param="imu0_config">[false,  false, false,
                                       false, false, false,
                                       false, false, false,
                                       true , true, true,
                                       true, true, true]</rosparam>
        
        <!-- Lidar IMU config-->
        <param name="imu0" value="/ouster/imu"/>
        <param name="imu0_differential" value="false" />
        <param name="imu0_relative" value="true" />
        <param name="imu0_queue_size" value="10" />
        <param name="imu0_remove_gravitational_acceleration" value="true" />
        <rosparam param="imu0_config">[false,  false, false,
                                       false, false, false,
                                       false, false, false,
                                       true , true, true,
                                       true, true, true]</rosparam>

        <param name="print_diagnostics" value="true" />
        <param name="debug" value="false" />
        <param name="debug_out_file" value="debug_odom_ekf.txt" />
        <rosparam param="process_noise_covariance">[0.05, 0,    0, 0, 0, 0, 0, 0, 0, 0, 0, 0,    0, 0, 0,

                                                    0, 0.05, 0, 0, 0, 0, 0, 0, 0,  0, 0, 0, 0, 0, 0,

                                                    0, 0, 0.06, 0, 0, 0, 0, 0, 0,  0, 0, 0, 0, 0, 0,

                                                    0, 0, 0, 0.03, 0, 0, 0, 0, 0,  0, 0, 0, 0, 0, 0,

                                                    0, 0, 0, 0, 0.03, 0, 0, 0, 0,  0, 0, 0, 0, 0, 0,

                                                    0, 0, 0, 0, 0, 0.06, 0, 0, 0,  0, 0, 0, 0, 0, 0,

                                                    0, 0, 0, 0, 0, 0, 0.025,0, 0,  0, 0, 0, 0, 0, 0,

                                                    0, 0, 0, 0, 0, 0, 0, 0.025,0,  0, 0, 0, 0, 0, 0,

                                                    0, 0, 0, 0, 0, 0, 0, 0, 0.04, 0,    0, 0, 0, 0, 0,

                                                    0, 0, 0, 0, 0, 0, 0, 0, 0,  0.01, 0, 0, 0, 0, 0,

                                                    0, 0, 0, 0, 0, 0, 0, 0, 0,  0, 0.01, 0, 0, 0, 0,

                                                    0, 0, 0, 0, 0, 0, 0, 0, 0,  0, 0, 0.02, 0, 0, 0,

                                                    0, 0, 0, 0, 0, 0, 0, 0, 0,  0, 0, 0, 0.01, 0, 0,

                                                    0, 0, 0, 0, 0, 0, 0, 0, 0,  0, 0, 0, 0, 0.01, 0,

                                                    0, 0, 0, 0, 0, 0, 0, 0, 0,  0, 0, 0, 0, 0, 0.005]</rosparam>

        <rosparam param="initial_estimate_covariance">[1e-9, 0,    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,    0, 0,

                                                        0, 1e-9, 0, 0, 0, 0, 0, 0, 0, 0,    0, 0, 0, 0, 0,

                                                        0, 0, 1e-9, 0, 0, 0, 0, 0, 0, 0,    0, 0, 0, 0, 0,

                                                        0, 0, 0, 1e-9, 0, 0, 0, 0, 0, 0,    0, 0, 0, 0, 0,

                                                        0, 0, 0, 0, 1e-9, 0, 0, 0, 0, 0,    0, 0, 0, 0, 0,

                                                        0, 0, 0, 0, 0, 1e-9, 0, 0, 0, 0,    0, 0, 0, 0, 0,

                                                        0, 0, 0, 0, 0, 0, 1 , 0, 0, 0,    0, 0, 0, 0, 0,

                                                        0, 0, 0, 0, 0, 0, 0, 1 , 0, 0,    0, 0, 0, 0, 0,

                                                        0, 0, 0, 0, 0, 0, 0, 0, 1e-9, 0,    0, 0, 0, 0, 0,

                                                        0, 0, 0, 0, 0, 0, 0, 0, 0, 1,    0, 0, 0, 0, 0,

                                                        0, 0, 0, 0, 0, 0, 0, 0, 0, 0,    1, 0, 0, 0, 0,

                                                        0, 0, 0, 0, 0, 0, 0, 0, 0, 0,    0, 1 , 0, 0, 0,

                                                        0, 0, 0, 0, 0, 0, 0, 0, 0, 0,    0, 0, 1, 0, 0,

                                                        0, 0, 0, 0, 0, 0, 0, 0, 0, 0,    0, 0, 0, 1e-9, 0,

                                                        0, 0, 0, 0, 0, 0, 0, 0, 0, 0,    0, 0, 0, 0, 1e-9]</rosparam>

    </node>

    <node pkg="robot_localization" type="navsat_transform_node" name="navsat_transform_node" respawn="true">

        <param name="magnetic_declination_radians" value="0"/>
        <param name="yaw_offset" value="0"/>
        <param name="publish_filtered_gps" value="true"/>
        <param name="broadcast_utm_transform" value="true"/>

        <remap from="/imu/data" to="/imu" />
        <remap from="/gps/fix" to="/navsat/fix" />
        <remap from="/odometry/filtered" to="/odom/ekf/enc_imu" />

    </node>
</launch>